include/RAJA/policy/cuda/MemUtils_CUDA.hpp

```diff
@@ -275,54 +281,94 @@ bool setupReducers() { return detail::tl_status.setup_reducers; }
+
+//! get dynamic shared memory usage for current launch
+RAJA_INLINE
+size_t currentDynamicShmem() { return *detail::tl_status.dynamic_smem; }
+
+//! get maximum dynamic shared memory for current launch
+RAJA_INLINE
+size_t maxDynamicShmem()
+{
+  cudaFuncAttributes func_attr;
+  cudaErrchk(cudaFuncGetAttributes(&func_attr, detail::tl_status.func));
+  return func_attr.maxDynamicSharedSizeBytes;
+}
+
+constexpr size_t dynamic_smem_allocation_failure = std::numeric_limits<size_t>::max();
+
+//! Allocate dynamic shared memory for current launch
+//
+//  The first argument is a functional object that takes the maximum number of
+//  objects that can fit into the dynamic shared memory available and returns
+//  the number of objects to allocate.
+//  The second argument is the required alignment.
+//
+//  Returns an offset into dynamic shared memory aligned to align on success,
+//  or dynamic_smem_allocation_failure on failure. Note that asking for 0 memory
+//  takes the failure return path.
+template < typename T, typename GetNFromMax >
+RAJA_INLINE
+size_t allocateDynamicShmem(GetNFromMax&& get_n_from_max, size_t align = alignof(T))
+{
+  const size_t unaligned_shmem = *detail::tl_status.dynamic_smem;
+  const size_t align_offset = ((unaligned_shmem % align) != size_t(0))
+      ? align - (unaligned_shmem % align)
+      : size_t(0);
+  const size_t aligned_shmem = unaligned_shmem + align_offset;
+
+  const size_t max_shmem_bytes = maxDynamicShmem() - aligned_shmem;
+  const size_t n_bytes = sizeof(T) *
+      std::forward<GetNFromMax>(get_n_from_max)(max_shmem_bytes / sizeof(T));
+
+  if (size_t(0) < n_bytes && n_bytes <= max_shmem_bytes) {
+    *detail::tl_status.dynamic_smem = aligned_shmem + n_bytes;
+    return aligned_shmem;
+  } else {
+    return dynamic_smem_allocation_failure;
+  }
+}
+
 //! get resource for current launch
 RAJA_INLINE
 ::RAJA::resources::Cuda currentResource() { return detail::tl_status.res; }
 
 //! create copy of loop_body that is setup for device execution
+//
+// Note: This is done to setup the Reducer and MultiReducer objects through
+// their copy constructors. Both look at tl_status to setup per kernel launch
+// resources.
 template <typename LOOP_BODY>
 RAJA_INLINE typename std::remove_reference<LOOP_BODY>::type make_launch_body(
+    const void* func,
     cuda_dim_t gridDim,
     cuda_dim_t blockDim,
-    size_t RAJA_UNUSED_ARG(dynamic_smem),
+    size_t& dynamic_smem,
     ::RAJA::resources::Cuda res,
     LOOP_BODY&& loop_body)
 {
-  detail::SetterResetter<bool> setup_reducers_srer(
-      detail::tl_status.setup_reducers, true);
-  detail::SetterResetter<::RAJA::resources::Cuda> res_srer(
-      detail::tl_status.res, res);
-
-  detail::tl_status.gridDim = gridDim;
-  detail::tl_status.blockDim = blockDim;
+  ::RAJA::detail::ScopedAssignment<detail::cudaInfo> info_sa(detail::tl_status,
+      detail::cudaInfo{func, gridDim, blockDim, &dynamic_smem, res, true});
```

Comment:
If I understand this correctly, the CUDA setup that had previously occurred here is now relegated to the copy constructors of the Reducer objects, and this is why you need `cudaInfo` to have the kernel func? And, this is how both the MultiReducer and the original Reducer will now operate?

Reply:
Ya, this just consolidates all of the changes instead of having them be separate. Ya this changes the thread local state that is read in the copy constructors of he Reducer and MultiReducer objects to do the launch or per-kernel setup.